### PR TITLE
Equip json_arrayt and json_objectt with a full API to enable hiding [blocks: #3602]

### PR DESCRIPTION
**Pull request by tautschnig**
As the comment said, various members of jsont should not need to be public.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 129c4a01311f3bfa90ca3f44eef63b54d085d5c6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95280964
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by tautschnig** [comment]
@peterschrammel @thk123 @allredj It is perfectly likely that this needs a genuine change in TG, could you please check?

**Comment by allredj** [comment]
@tautschnig yes it does. Creating a TG update now.

**Comment by allredj** [comment]
Bump is here: https://github.com/diffblue/test-gen/pull/2579.

**Comment by allredj** [comment]
TG bump passing

**Comment by tautschnig** [comment]
Thanks @allredj for the immediate action! Assigning to more people for a review.

**Review by peterschrammel — APPROVED** [review]

File: src/goto-analyzer/taint_parser.cpp [hunk 1]
```diff
@@ -40,28 +40,24 @@ bool taint_parser(
     return true;
   }
 
-  for(jsont::arrayt::const_iterator
-      it=json.array.begin();
-      it!=json.array.end();
-      it++)
+  for(const auto &j : to_json_array(json))
```
Comment (smowton) [inline src/goto-analyzer/taint_parser.cpp]: `j` -> `taint_spec`?

File: src/goto-cc/linker_script_merge.cpp [hunk 2]
```diff
@@ -738,38 +740,50 @@ int linker_script_merget::goto_and_object_mismatch(
...
-                 j.object.find("size") != j.object.end() &&
-                 j.object.find("annot") != j.object.end() &&
-                 j.object.find("commt") != j.object.end() &&
-                 j.object.find("start-symbol") != j.object.end() &&
-                 j.object.find("has-end-symbol") != j.object.end() &&
-                 j.object.find("section") != j.object.end() &&
+          if(!j.is_object())
+            return false;
+
+          const json_objectt &j_object = to_json_object(j);
```
Comment (smowton) [inline src/goto-cc/linker_script_merge.cpp]: `j_object` -> `region`?

**Review by smowton — APPROVED** [review]
suggest changing the various `j` and `j_object` names to something more informative, otherwise lgtm

**Comment by tautschnig** [comment]
@smowton Thanks for pushing me to it, renaming done.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 6a8d1319d49533a020451156b3ef143a178856f8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95393199
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by kroening — APPROVED** [review]

**Comment by tautschnig** [comment]
Assigning to @allredj @peterschrammel @thk123 to co-ordinate the merge of this PR with the merge of the corresponding TG bump in order to avoid an unhappy TG bot for an extended period of time.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 5167bcb818be2957e059e4362cd0fc422125c58c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96133175
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by allredj** [comment]
Will merge today.

**Pull request merged**